### ISSUE TITLE
Merge Datasets

### ISSUE THREAD

**Issue by joerg-halo** [halo-db/storymap#43]
A a Scientist I want to merge datasets from several instruments obtained from one research flight. Thereby, I want to run retrievals relying on several instruments. Moreover, I also want to compare retrievals relying on data from different instruments.

**Comment by bjoernbroetz** [comment]
Merging is already a scientific activity. You must align the time coordinate and maybe even adjust it to take care of different speed of the instrument clocks. So we must be careful here.

**Comment by d70-t** [comment]
For the specMACS database, we've built in derived datasets. The idea is as follows: there are file-based datasets as well as derived datasets. Each of those datasets has its own unique id. File-based datasets get their id by hashing the datasets contents. Derived datasets get their id by hashing the function which transforms a dataset as well as all the ids of its inputs. This way, each dataset has its own id (which could be transformed into a DOI) and independent of being file-based or derived, the dataset could always be accessed via its id (and OpenDAP).

The is a "database" involved (in this case MongoDB), which associates the dataset ids either to the filename of the source file or to the function which transforms the dataset and a list of input ids to the function.

An additional feature of this approach has been the possibility to back-propagate slices of the requested dataset. That is, if the user e.g. requests only a certain temporal slice, then the function may decide that only a certain temporal slice of the input data is required and thus only a part of the requested dataset is computed.

For derived datasets which are complicated to compute, a caching layer has been implemented, such that those datasets which are hard to compute but easy to store don't need to be recomputed again and again.

**Comment by bjoernbroetz** [comment]
The question of merging touches the **big** topic of processing at the HALO-DB. It implies potentially the subjects:
- Controlled Vocabulary/ Ontology naming convention
- Provenance
- Reproducibility of previous  processing (software changes) as @rico-hengst mentioned
- Processing time, queueing, delayed data delivery 
- Costs of processing
- Interactive adjustments to the data (time coordinate shift & stretch) by a scientific user
    - interactive visualisation
    - recording of processing steps
    - providing reproducible processing instructions (persistent and with an identifyer)
- Database for primary data (vs. metadata database)

**Comment by HeAloDi05** [comment]
I don't think that creating merged files is a scientific activity. I see it as an original service that a database can (though not necessarily must) provide. Time synchronization on HALO already happens during data recording (there is a common time standard on board the aircraft). The different time resolution of the instruments is not subject to post-processing of the data in the database. This has already been taken into account in the data given to the database. The common time axis is given by the beginning and end of a flight. Merged files are the prerequisite for further, then scientific work, with the data. In the past, merged files were formed with a coarser time resolution than the primary data files (i.e., 10 s averages). Discontinuous parameters may be adjusted to the merged time axis by interpolation.
